tempesta_fw/t/functional/tests/test_frang.py

```diff
@@ -0,0 +1,516 @@
+sys.path.append(dirname(realpath(__file__))+ sep + sep + "helpers")
+
+import conf
+import tfw
+from socket import *
+import select
+import struct
+
+global tcount
+
+tcount = 0
+
+class Test:
+	def __init__(self):
+		self.vs_get = b"GET / HTTP/1.0\r\nhost: loc\r\n\r\n"
+		self.s = socket(AF_INET, SOCK_STREAM)
+		self.cfg = conf.Config('etc/tempesta_fw.conf')
+		self.cfg.add_option('cache', '0')
+		self.cfg.add_option('listen', '8081')
+		self.cfg.add_option('server', '127.0.0.1:80')
+
+	def uri_len(self):
+		global tcount
+
+		self.res = False
+		print("uri\n")
+		self.__init__()
+		self.cfg.add_section('frang_limits')
+		self.cfg.add_option('http_uri_len', '3')
+		self.cfg.add_end_of_section()
+		self.vs_get = b"GET /index.html HTTP/1.0\r\nhost: loc\r\n\r\n"
+		tfw.start_with_frang()
+		self.s.connect(("127.0.0.1",8081))
+		print("tfw start\n")
+		try:
+			self.s.send(self.vs_get)
+			data = self.s.recv(1024)
+			if len(data) == 0:
+				self.res = True
+				tcount += 1
+		except OSError as e:
+			print(e)
+		print("res:", self.res)
+		time.sleep(5)
+		tfw.stop()
+
+	def request_rate(self):
+		global tcount
+
+		self.res = False
+		print("req_rate\n")
+		self.__init__()
+		self.cfg.add_section('frang_limits')
+		self.cfg.add_option('request_rate', '5')
+
+		self.cfg.add_end_of_section()
+		tfw.start_with_frang()
+		print("tfw start\n")
+		self.s.connect(("127.0.0.1",8081))
+
+		self.vs_get = b"GET / HTTP/1.0\r\nhost: loc\r\n" +\
+b"Connection: Keep-Alive\r\n\r\n"
+		try:
+			for x in range(0, 15):
+				self.s.sendall(self.vs_get)
+
+		except OSError as e:
+			self.res = True
+			tcount += 1
+
+		self.s.close()
+		print("res:", self.res)
+		time.sleep(5)
+		tfw.stop()
+
+	def request_burst(self):
+		global tcount
+
+		self.res = False
+		print("req_burst\n")
+		self.__init__()
+		self.cfg.add_section('frang_limits')
+		self.cfg.add_option('request_burst', '1')
+
+		self.cfg.add_end_of_section()
+		tfw.start_with_frang()
+		print("tfw start\n")
+		self.s.connect(("127.0.0.1",8081))
+
+		self.vs_get = b"GET / HTTP/1.0\r\nhost: loc\r\n" +\
+b"Connection: Keep-Alive\r\n\r\n"
+		try:
+			for x in range(0, 15):
+				self.s.sendall(self.vs_get)
+
+		except OSError as e:
+			self.res = True
+			tcount += 1
+
+		self.s.close()
+		print("res:", self.res)
+		time.sleep(5)
+		tfw.stop()
+
+	def conn_max(self):
+		global tcount
+		self.res = False
+		print("conn_max\n")
+		self.__init__()
+		self.cfg.add_section('frang_limits')
+		self.cfg.add_option('ip_block', 'on')
+		self.cfg.add_option('concurrent_connections', '5')
+		self.cfg.add_end_of_section()
+		self.cfg.add_option('keepalive_timeout', '3')
+		tfw.start_with_frang()
+		print("tfw start\n")
+		self.vs_get = b"GET / HTTP/1.0\r\nhost: loc\r\n" +\
+b"Connection: Keep-Alive\r\n\r\n"
+
+		try:
+			socks = []
+			conncount = 0
+			port = 8095
+			for x in range(0,7):
+				s = socket(AF_INET, SOCK_STREAM)
+				s.settimeout(2)
+				s.setsockopt(SOL_SOCKET, SO_REUSEADDR, 1)
+				s.bind(('127.0.0.6', port))
```

Comment:
The logic seems broken. There is no sense to bind all the connection to the same port. I reckon you simply get exceptions here since 7 connection can not live with the same addresses and ports concurrently.